File: apk/app/src/main/java/com/example/momentra/ui/setup/SetupChipGrid.kt
package com.example.momentra.ui.setup

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

data class SetupChipOption(
    val value: String,
    val label: String,
    val emoji: String? = null,
)

@Composable
fun SetupChipGrid(
    options: List<SetupChipOption>,
    selected: Set<String>,
    onToggle: (String) -> Unit,
    columns: Int = 2,
    selectedColor: Color = SetupTokens.ChipSelected,
    unselectedColor: Color = SetupTokens.ChipUnselected,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        options.chunked(columns).forEach { row ->
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                row.forEach { option ->
                    SetupChipCell(
                        option = option,
                        isSelected = option.value in selected,
                        onToggle = onToggle,
                        selectedColor = selectedColor,
                        unselectedColor = unselectedColor,
                        modifier = Modifier.weight(1f),
                    )
                }
                repeat(columns - row.size) {
                    androidx.compose.foundation.layout.Spacer(modifier = Modifier.weight(1f))
                }
            }
        }
    }
}

@Composable
private fun SetupChipCell(
    option: SetupChipOption,
    isSelected: Boolean,
    onToggle: (String) -> Unit,
    selectedColor: Color,
    unselectedColor: Color,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier = modifier
            .clip(RoundedCornerShape(12.dp))
            .background(if (isSelected) selectedColor else unselectedColor)
            .border(
                width = 1.dp,
                color = if (isSelected) selectedColor else SetupTokens.BorderSubtle,
                shape = RoundedCornerShape(12.dp),
            )
            .clickable { onToggle(option.value) }
            .padding(horizontal = 12.dp, vertical = 14.dp)
            .semantics {
                role = Role.Checkbox
                contentDescription = "${option.label}, ${if (isSelected) "selected" else "not selected"}"
            },
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        if (option.emoji != null) {
            Text(option.emoji, fontSize = 18.sp)
        }
        Text(
            option.label,
            color = SetupTokens.TextPrimary,
            fontSize = 13.sp,
            fontWeight = if (isSelected) FontWeight.SemiBold else FontWeight.Normal,
        )
    }
}

@Composable
fun SetupChipRow(
    options: List<SetupChipOption>,
    selected: String?,
    onSelect: (String) -> Unit,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier = modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        options.forEach { option ->
            val isSelected = option.value == selected
            Text(
                text = buildString {
                    if (option.emoji != null) append("${option.emoji} ")
                    append(option.label)
                },
                color = SetupTokens.TextPrimary,
                fontSize = 13.sp,
                fontWeight = if (isSelected) FontWeight.SemiBold else FontWeight.Normal,
                modifier = Modifier
                    .weight(1f)
                    .clip(RoundedCornerShape(12.dp))
                    .background(if (isSelected) SetupTokens.ChipSelected else SetupTokens.ChipUnselected)
                    .border(
                        1.dp,
                        if (isSelected) SetupTokens.ChipSelected else SetupTokens.BorderSubtle,
                        RoundedCornerShape(12.dp),
                    )
                    .clickable { onSelect(option.value) }
                    .padding(vertical = 12.dp, horizontal = 8.dp)
                    .semantics {
                        role = Role.RadioButton
                        contentDescription = option.label
                    },
            )
        }
    }
}

File: momentra/momentra/Shell/Setup/SetupChipGrid.swift
import SwiftUI

struct SetupChipOption: Identifiable, Hashable {
    let value: String
    let label: String
    var emoji: String? = nil

    var id: String { value }
}

struct SetupChipGrid: View {
    let options: [SetupChipOption]
    let selected: Set<String>
    let onToggle: (String) -> Void
    var columns: Int = 2
    var selectedColor: Color = SetupTokens.chipSelected
    var unselectedColor: Color = SetupTokens.chipUnselected

    var body: some View {
        VStack(spacing: 8) {
            ForEach(Array(options.chunked(into: columns).enumerated()), id: \.offset) { _, row in
                HStack(spacing: 8) {
                    ForEach(row) { option in
                        SetupChipCell(
                            option: option,
                            isSelected: selected.contains(option.value),
                            onToggle: onToggle,
                            selectedColor: selectedColor,
                            unselectedColor: unselectedColor
                        )
                    }
                    if row.count < columns {
                        ForEach(0..<(columns - row.count), id: \.self) { _ in
                            Color.clear.frame(maxWidth: .infinity)
                        }
                    }
                }
            }
        }
    }
}

private struct SetupChipCell: View {
    let option: SetupChipOption
    let isSelected: Bool
    let onToggle: (String) -> Void
    let selectedColor: Color
    let unselectedColor: Color

    var body: some View {
        Button {
            onToggle(option.value)
        } label: {
            HStack(spacing: 8) {
                if let emoji = option.emoji {
                    Text(emoji).font(.system(size: 18))
                }
                Text(option.label)
                    .font(.system(size: 13, weight: isSelected ? .semibold : .regular))
                    .foregroundStyle(SetupTokens.textPrimary)
                    .multilineTextAlignment(.leading)
                    .frame(maxWidth: .infinity, alignment: .leading)
            }
            .padding(.horizontal, 12)
            .padding(.vertical, 14)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(isSelected ? selectedColor : unselectedColor, in: RoundedRectangle(cornerRadius: 12))
            .overlay(
                RoundedRectangle(cornerRadius: 12)
                    .stroke(isSelected ? selectedColor : SetupTokens.borderSubtle, lineWidth: 1)
            )
        }
        .buttonStyle(.plain)
        .accessibilityLabel("\(option.label), \(isSelected ? "selected" : "not selected")")
    }
}

struct SetupChipRow: View {
    let options: [SetupChipOption]
    let selected: String?
    let onSelect: (String) -> Void

    var body: some View {
        HStack(spacing: 8) {
            ForEach(options) { option in
                let isSelected = option.value == selected
                Button {
                    onSelect(option.value)
                } label: {
                    Text([option.emoji, option.label].compactMap { $0 }.joined(separator: " "))
                        .font(.system(size: 13, weight: isSelected ? .semibold : .regular))
                        .foregroundStyle(SetupTokens.textPrimary)
                        .frame(maxWidth: .infinity)
                        .padding(.vertical, 12)
                        .padding(.horizontal, 8)
                        .background(isSelected ? SetupTokens.chipSelected : SetupTokens.chipUnselected, in: RoundedRectangle(cornerRadius: 12))
                        .overlay(
                            RoundedRectangle(cornerRadius: 12)
                                .stroke(isSelected ? SetupTokens.chipSelected : SetupTokens.borderSubtle, lineWidth: 1)
                        )
                }
                .buttonStyle(.plain)
                .accessibilityLabel(option.label)
            }
        }
    }
}

private extension Array {
    func chunked(into size: Int) -> [[Element]] {
        guard size > 0 else { return [self] }
        return stride(from: 0, to: count, by: size).map {
            Array(self[$0..<Swift.min($0 + size, count)])
        }
    }
}
